src/NuGet.Core/NuGet.Protocol/Resources/VulnerabilityInfoResourceV3.cs

```diff
@@ -0,0 +1,299 @@
+
+            HttpSourceCacheContext httpSourceCacheContext = HttpSourceCacheContext.Create(cacheContext, isFirstAttempt: true);
+            var request = new HttpSourceCachedRequest(vulnerabilityPage.Url.OriginalString, "vuln_data_" + vulnerabilityPage.Name, httpSourceCacheContext);
+            IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>? data;
+            try
+            {
+                data = await httpSourceResource.HttpSource.GetAsync(request,
+                    result =>
+                    {
+                        using (var textReader = new StreamReader(result.Stream))
+                        using (var jsonReader = new JsonTextReader(textReader))
+                        {
+                            var parsed = JsonExtensions.JsonObjectSerializer.Deserialize<IgnoreCaseDictionary<IReadOnlyList<VulnerabilityInfo>>>(jsonReader);
+                            return Task.FromResult(parsed);
+                        }
+                    },
+                    logger,
+                    cancellationToken);
+            }
+            catch (Exception ex) when (ex is not FatalProtocolException)
+            {
+                // Deserialization errors might throw Newtonsoft.Json's JsonSerializationException, but the library we deserialize with
+                // should not be a leaky abstraction, so it should be just a little easier to maybe move to System.Text.Json one day.
+                throw new FatalProtocolException(message: ex.Message, innerException: ex);
+            }
+
+            if (data == null)
+            {
+                string message = "Package source '{0}' vulnerability data file '{1}' ({2}) did not successfully deserialize. This indicates a problem with the JSON file contents returned from the server.";
+                throw new FatalProtocolException(message);
+            }
+
+            return data;
+        }
+
+        /// <inheritdoc cref="IVulnerabilityInfoResource.GetVulnerabilityInfoAsync(SourceCacheContext, ILogger, CancellationToken)"/>
+        public async Task<GetVulnerabilityInfoResult> GetVulnerabilityInfoAsync(SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
+        {
+            List<Exception>? exceptions = null;
+            List<IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>>? knownVulnerabilities = null;
+
+            try
+            {
+                IReadOnlyList<V3VulnerabilityIndexEntry> indexEntries = await GetVulnerabilityFilesAsync(cacheContext, logger, cancellationToken);
+
+                if (indexEntries.Count == 0)
+                {
+                    knownVulnerabilities = new();
+                    return new GetVulnerabilityInfoResult(knownVulnerabilities, ToAggregateException(exceptions));
+                }
+
+                var tasks = new Task<IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>>[indexEntries.Count];
```

Comment:
How do we account for an data file being already on disk an up to date?
I'm not sure I see that handled anywhere. Aren't we re-downloading the files constantly now?

Reply:
It will use the cache settings on the passed-in `SourceCacheContext`. This is a perf improvement, but not a feature correctness issue, so will be implemented later when there aren't more pressing things to work on: https://github.com/NuGet/Client.Engineering/issues/2244